### PR TITLE
Monorepo

### PR DESCRIPTION
**Pull request by VictoriaBeilsten-Edmands**
Restructure component and example in client and use pnpm packages.
Update dependencies

### CODE REVIEW

File: client/component/src/AnyPlot.tsx [hunk 1]
```diff
@@ -9,25 +10,16 @@ import SurfacePlot from './SurfacePlot';
...
-  | Int8Array
-  | Int16Array
-  | Int32Array
-  | Uint8Array
-  | Uint8ClampedArray
-  | Uint16Array
-  | Uint32Array
-  | Float32Array
-  | Float64Array;
 type NdArray<T extends GenericArray<unknown>> = import('ndarray').NdArray<T>;
```
Comment (PeterC-DLS) [inline client/component/src/AnyPlot.tsx:23]: Update this too and remove its export

File: client/component/vite.config.ts [hunk 2]
```diff
@@ -1,20 +1,16 @@
...
 import dts from 'vite-plugin-dts';
 import { resolve } from 'path';
 
 // https://vitejs.dev/config/
 export default defineConfig({
-  plugins: [
-    react(),
-    dts({ rollupTypes: true }),
-  ],
+  plugins: [react(), dts({ rollupTypes: true })] as PluginOption[],
```
Comment (PeterC-DLS) [inline client/component/vite.config.ts:8]: Is this necessary after updating the package dependencies and reinstalling?

File: tsconfig.json [hunk 3]
```diff
@@ -2,22 +2,32 @@
...
     "jsx": "react-jsx",
-
     /* Linting */
     "strict": true,
     "noUnusedLocals": true,
     "noUnusedParameters": true,
     "noFallthroughCasesInSwitch": true
   },
+  "include": [
+    "client/**/*.ts",
```
Comment (PeterC-DLS) [inline tsconfig.json:26]: Just include client

File: client/component/src/AnyPlot.tsx [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  ColorScaleType,
+  CustomDomain,
+  DAxesParameters,
+  DLineData,
+  Domain,
+  HeatmapPlotProps,
+  ImagePlotProps,
+  LinePlotProps,
+  // eslint-disable-next-line react-refresh/only-export-components
+  MP_NDArray,
```
[the remark this answers is not in the document]
Comment (VictoriaBeilsten-Edmands) [reply client/component/src/AnyPlot.tsx:179]: I exported it to use in utils.ts

File: client/component/src/AnyPlot.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import TableDisplay from './TableDisplay';
+import { measureInteraction } from './utils';
+import type { SelectionBase } from './selections/utils';
+
+type Aspect = import('@h5web/lib').Aspect;
+type AxisScaleType = import('@h5web/lib').AxisScaleType;
+type ColorMap = import('@h5web/lib').ColorMap;
+type ColorScaleType = import('@h5web/lib').ColorScaleType;
+type CustomDomain = import('@h5web/lib').CustomDomain;
+type Domain = import('@h5web/lib').Domain;
```
[the remark this answers is not in the document]
Comment (VictoriaBeilsten-Edmands) [reply client/component/src/AnyPlot.tsx]: I've removed these. Only `type NdArray<T extends GenericArray<unknown>> = import('ndarray').NdArray<T>;` is left.

File: client/component/src/AnyPlot.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (PeterC-DLS) [reply client/component/src/AnyPlot.tsx:179]: I meant NdArray (the file changed during my review so this comment got misplaced).

File: package.json [hunk 6]
```diff
@@ -3,68 +3,37 @@
...
+    "eslint-plugin-react": "^7.33.2",
     "eslint-plugin-react-hooks": "^4.6.0",
-    "prettier": "^2.8.2",
-    "typescript": "^5.1.3",
-    "vite": "^4.5.0",
+    "prettier": "^2.8.8",
+    "typescript": "^5.3.2",
+    "vite": "^4.5.1",
     "vitest": "^0.34.6",
     "watch": "^1.0.2"
```
[the remark this answers is not in the document]
Comment (PeterC-DLS) [reply package.json:32]: Vite in dev mode watches for changes already

File: client/component/vite.config.ts (hunk 2 above)
Comment (VictoriaBeilsten-Edmands) [reply client/component/vite.config.ts]: The `PluginOption `part?

Comment (PeterC-DLS) [reply client/component/vite.config.ts]: Yes, it looks like it's gone now.

File: .eslintrc.cjs [hunk 7]
```diff
@@ -9,7 +9,7 @@ module.exports = {
     'plugin:react-hooks/recommended',
     'prettier',
   ],
-  ignorePatterns: ['*.css'],
+  ignorePatterns: ['**/*.css', '**/*.json', '**/*.d.ts', '**/vite.config.ts'],
```
Comment (PeterC-DLS) [inline .eslintrc.cjs:12]: Combine all the ignorePatterns here from lower configurations

File: client/component/vite-env.d.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+/// <reference types="vite/client" />
```
Comment (PeterC-DLS) [inline client/component/vite-env.d.ts:1]: Can remove this file as you only need one declaration file stating this

File: client/example/src/vite-env.d.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+/// <reference types="vite/client" />
```
Comment (PeterC-DLS) [inline client/example/src/vite-env.d.ts:1]: Can remove this

File: package.json [hunk 10]
```diff
@@ -3,70 +3,32 @@
...
-    "start": "vite",
+    "start": "pnpm --filter {client/example} start",
     "build": "vite build",
-    "preview": "vite preview",
+    "packages": "pnpm --filter {client/*} build",
+    "start:storybook": "pnpm --filter storybook start",
+    "build:storybook": "pnpm --filter storybook build",
     "test": "vitest",
     "prettify": "prettier -w client/**",
     "lint": "eslint \"client/**\" --ext .ts --ext .tsx",
```
Comment (PeterC-DLS) [inline package.json:15]: Can simplify these scripts commands as you have set the ignore patterns. Also rationalize so these top level scripts call the scripts in lower level package.jsons

File: package.json [hunk 11]
```diff
@@ -3,70 +3,32 @@
   "name": "davidia",
   "version": "0.1.0",
   "private": true,
-  "type": "module",
+  "packageManager": "pnpm@8.6.0",
   "scripts": {
-    "preinstall": "npx only-allow pnpm",
-    "start": "vite",
+    "start": "pnpm --filter {client/example} start",
     "build": "vite build",
```
Comment (PeterC-DLS) [inline package.json:9]: This needs to build both sub projects

**Comment by PeterC-DLS** [comment]
Also need to fix the server startup with
```
--- a/davidia/main.py
+++ b/davidia/main.py
@@ -115,6 +115,7 @@ def add_benchmark_endpoint():
 def add_client_endpoint(client_path):
     logging.debug("Adding /client endpoint which uses %s", client_path)
     app.mount("/client", StaticFiles(directory=client_path, html=True), name="webui")
+    app.mount("/assets", StaticFiles(directory=client_path / "assets", html=True), name="webui")
 
 
 def create_parser():
@@ -125,7 +126,7 @@ def create_parser():
         "-b", "--benchmark", help="Add /benchmark endpoint", action="store_true"
     )
     parser.add_argument(
-        "-c", "--client", help="Add /client endpoint using given directory (or 'dist')", nargs='?', const='dist', default=SUPPRESS
+        "-c", "--client", help="Add /client endpoint using given directory (or 'dist')", nargs='?', const='client/example/dist', default=SUPPRESS
     )
     return parser
```

File: package.json [hunk 12]
```diff
@@ -6,14 +6,13 @@
   "packageManager": "pnpm@8.6.0",
   "scripts": {
     "start": "pnpm --filter {client/example} start",
-    "build": "vite build",
     "packages": "pnpm --filter {client/*} build",
```
Comment (PeterC-DLS) [inline package.json:9]: Rename this to build

File: .eslintrc.cjs (hunk 7 above)
Comment (VictoriaBeilsten-Edmands) [reply .eslintrc.cjs]: Should have all now

**Review by PeterC-DLS — APPROVED** [review]
All done!

**Pull request merged**